File: day_08/solution.py
from dataclasses import dataclass

@dataclass
class Cell:
    x: int
    y: int
    height: int
    vis_from_left: bool = False
    vis_from_right: bool = False
    vis_from_top: bool = False
    vis_from_bottom: bool = False
    area: int = 0



def look_right(cell):
    i = 0
    for i in range(1, len(grid[cell.y])-cell.x):
        print(i)
        if grid[cell.y][cell.x+i].height >= cell.height:
            break
    return i


def look_left(cell):
    i = 0
    for i in range(1, cell.x+1):
        if grid[cell.y][cell.x-i].height >= cell.height:
            break
    return i

def look_up(cell):
    i = 0
    for i in range(1, cell.y+1):
        if grid[cell.y-i][cell.x].height >= cell.height:
            break
    return i

def look_down(cell):
    i = 0
    for i in range(1, len(grid)-cell.y):
        if grid[cell.y+i][cell.x].height >= cell.height:
            break
    return i


def calc_area(cell):
    return look_right(cell) * look_left(cell) * look_up(cell) * look_down(cell)



with open('./day_08/input.txt') as f:
    big_string = f.read()



grid = []

for i, line in enumerate(big_string.splitlines()):
    grid.append([])
    for j, char in enumerate(line):
        grid[i].append(Cell(j, i, int(char)))


# visible from left:
for i, row in enumerate(grid):
    max_height = 0
    for j, cell in enumerate(row):
        if cell.height > max_height or i == 0 or j == 0:
            cell.vis_from_left = True
            max_height = cell.height

# visible from right:
for i, row in enumerate(grid):
    max_height = 0
    for j, cell in enumerate(row[::-1]):
        if cell.height > max_height or i == 0 or j == 0:
            cell.vis_from_right = True
            max_height = cell.height

# visible from top:
for j, col in enumerate(zip(*grid)):
    max_height = 0
    for i, cell in enumerate(col):
        if cell.height > max_height or i == 0 or j == 0:
            cell.vis_from_top = True
            max_height = cell.height

# visible from bottom:
for j, col in enumerate(zip(*grid)):
    max_height = 0
    for i, cell in enumerate(col[::-1]):
        if cell.height > max_height or i == 0 or j == 0:
            cell.vis_from_bottom = True
            max_height = cell.height


# count visible cells:
visible_cells = 0
for row in grid:
    for cell in row:
        cell.area = calc_area(cell)
        if cell.vis_from_left or cell.vis_from_right or cell.vis_from_top or cell.vis_from_bottom:
            visible_cells += 1




#find cell with max area:
max_area = 0
for row in grid:
    for cell in row:
        if cell.area > max_area:
            print(cell)
            max_area = cell.area
            



print(visible_cells)
print(max_area)

File: day_08/solution.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Cell struct {
	x, y int
	height int
	vis_from_left bool 
	vis_from_right bool
	vis_from_top bool
	vis_from_bottom bool
	visible_area_score int
}

func NewCell(x, y, height int) *Cell {
	return &Cell{x, y, height, false, false, false, false, 0}
}

func (c *Cell) ViewToRight(grid [][]Cell) int {
	dist := 0
	for i := 1; i < len(grid[c.y])-c.x; i++ {
		dist = i
		if grid[c.y][c.x+i].height >= c.height {
			break
		}
	}
	return dist
}

func (c *Cell) ViewToLeft(grid [][]Cell) int {
	dist := 0
	for i := 1; i < c.x+1; i++ {
		dist = i
		if grid[c.y][c.x-i].height >= c.height {
			break
		}
	}
	return dist
}

func (c *Cell) ViewToTop(grid [][]Cell) int {
	dist := 0
	for i := 1; i < c.y+1; i++ {
		dist = i
		if grid[c.y-i][c.x].height >= c.height {
			break
		}
	}
	return dist
}

func (c *Cell) ViewToBottom(grid [][]Cell) int {
	dist := 0
	for i := 1; i < len(grid)-c.y; i++ {
		dist = i
		if grid[c.y+i][c.x].height >= c.height {
			break
		}
	}
	return dist
}

func (c *Cell) VisibleArea(grid [][]Cell) int {
	return c.ViewToRight(grid) * c.ViewToLeft(grid) * c.ViewToTop(grid) * c.ViewToBottom(grid)
}


func main() {
	start_time := time.Now()
	grid := [][]Cell{}

	file, err := os.ReadFile("input.txt"); if err != nil { panic(err) }
	lines := strings.Split(string(file), "\n")
	for y, line := range lines {
		row := []Cell{}
		for x, char := range line {
			height, err := strconv.Atoi(string(char)); if err != nil { panic(err) }
			newCell := NewCell(x, y, height)
			row = append(row, *newCell)
		}
		grid = append(grid, row)
	} 

	/* visible from left */
	for y, row := range grid {
		max_height := 0
		for x, cell := range row {
			if cell.height > max_height || x == 0 {
				grid[y][x].vis_from_left = true
				max_height = cell.height
			}
		}
	}
	
	/* visible from right */
	for _, row := range grid {
		max_height := 0
		for x := len(row)-1; x >= 0; x-- {
			if row[x].height > max_height || x == len(row)-1 {
				row[x].vis_from_right = true
				max_height = row[x].height
			}
		}
	}
	
	/* visible from top */
	for x := 0; x < len(grid[0]); x++ {
		max_height := 0
		for y := 0; y < len(grid); y++ {
			if grid[y][x].height > max_height || y == 0 {
				grid[y][x].vis_from_top = true
				max_height = grid[y][x].height
			}
		}
	}

	/* visible from bottom */
	for x := 0; x < len(grid[len(grid)-1]); x++ {
		max_height := 0
		for y := len(grid)-1; y >= 0; y-- {
			if grid[y][x].height > max_height || y == len(grid)-1 {
				grid[y][x].vis_from_bottom = true
				max_height = grid[y][x].height
			}
		}
	}

	/* count visible cells and find cell with max visibility */
	visible_cells := 0
	max_visible_area_score := 0
	for _, row := range grid {
		for _, cell := range row {
			cell.visible_area_score = cell.VisibleArea(grid)
			if cell.vis_from_left || cell.vis_from_right || cell.vis_from_top || cell.vis_from_bottom {
				visible_cells++
			}
			if cell.visible_area_score > max_visible_area_score {
				max_visible_area_score = cell.visible_area_score
			}
		}
	}
	fmt.Printf("Part 1 answer: %d\n", visible_cells)
	fmt.Printf("Part 2 answer: %d\n", max_visible_area_score)
	elapsed := time.Since(start_time)
	fmt.Printf("Time: %s\n", elapsed)
}